lib/kafka_gen_stage/consumer_logic.ex

```diff
@@ -28,71 +27,78 @@ defmodule KafkaGenStage.ConsumerLogic do
   @spec prepare_dispatch(
           buffered_msgs :: :queue.queue(),
           buffered_demand :: non_neg_integer(),
-          transformer :: (msg_tuple(), term() -> {term(), term()}),
-          transformer_state :: term,
-          bulk_transformer :: ([msg_tuple()] -> [msg_tuple()])
+          bulk_transformer :: ([msg_tuple()] -> [msg_tuple()]),
+          high_wm :: non_neg_integer()
         ) ::
           dispatch()
-  def prepare_dispatch(queue, 0, _transformer, state, _bulk_transformer) do
-    {[], :no_ack, 0, queue, state}
+  def prepare_dispatch(queue, 0, _bulk_transformer, _high_wm) do
+    {[], :no_ack, 0, queue}
   end
 
-  def prepare_dispatch(queue, demand, transformer, state, bulk_transformer) do
-    case :queue.out(queue) do
-      {{:value, msg}, queue} ->
-        {to_send, new_state, to_ack} = transform(msg, transformer, state)
-        new_demand = lower_demand(demand, to_send)
-
-        {to_send, to_ack, _demand, queue, new_transformer_state} =
-          prepare_dispatch(queue, new_demand, Enum.reverse(to_send), to_ack, transformer, new_state)
-
-        new_to_send = transform_bulk(to_send, bulk_transformer)
-        final_demand = lower_demand(demand, new_to_send)
-        {new_to_send, to_ack, final_demand, queue, new_transformer_state}
+  def prepare_dispatch(queue, demand, bulk_transformer, high_wm) do
+    {messages, ack_offset, queue} =
+      case :queue.out(queue) do
+        {{:value, {offset, _, _, _} = msg}, queue} ->
+          {to_send, to_ack, _demand, queue} =
+            prepare_dispatch(
+              queue,
+              demand - 1,
+              [msg],
+              offset,
+              high_wm
+            )
+
+          new_to_send =
+            to_send
+            |> Enum.reverse()
+            |> transform_bulk(bulk_transformer, high_wm)
+
+          {new_to_send, to_ack, queue}
+
+        {:empty, queue} ->
+          {transform_bulk([], bulk_transformer, high_wm), :no_ack, queue}
+      end
```

Comment:
There seems to be some unnecessary code duplication in prepare_dispatch/4 and prepare_dispatch/5. Why don't we unconditionally call prepare_dispatch/5 here? Something like this:

```
{to_send, to_ack, _demand, queue} = prepare_dispatch(queue, demand, [], :no_ack, high_wm)
to_send = … |> transform_bulk(…)
demand = lower_demand(demand, to_send)
{to_send, to_ack, demand, queue}
```

And we can also drop the unused `_demand` from prepare_dispatch/5 return value.

Reply:
I agree. I will do it.